### DISCUSSION TITLE
BodyManager::mActiveBodies

### DISCUSSION THREAD

**Discussion by feixinyv** [jrouwe/JoltPhysics#226] [Q&A] [answered]
Hello, I am trying to build a Prediction & Correction Roll-back Physics NetCode based on Jolt. Like "Rocket League" did ([https://www.gdcvault.com/play/1024972/It-IS-Rocket-Science-The](url)).
In BodyManager::SaveState, the code sorts a copy of mActiveBodies and save it. BodyManager::RestoreState uses the sorted copy to restore mActiveBodies, so the order of active Bodies might be different. 
If I save state at frame N, and keep going to Frame N+M, then restore state use Frame N's saved state and recalculate to Frame N+M. 
Whether the difference of active Bodies will cause deterministic problem?

**Comment by jrouwe** [comment, marked as answer]
Hello,

The order of bodies in mActiveBodies does not matter. Bodies can be activated from multiple threads so there is always non-determinism in there. To compensate for this the (contact) constraints are sorted during the physics simulation step. The only reason why SaveState / RestoreState sorts the active bodies is that the StateRecorder::IsValidating() mode can be on. In this case we're expecting SaveState to write bit for bit what is in the stream already. This is a convenient way of validating determinism: Write state 1, simulate, save state 2, restore state 1, simulate, save state in validation mode to check that the state is the same as 2.

W.r.t. saving the state at frame N and continuing to frame N + M, then rewinding to N and simulating again: This should give you the exact same results provided that you also call all API functions that modify physics state in the exact same frames. So if you did a Body::SetFriction on frame N + 2 then, when rewinding, you need to restore the friction to what is was on frame N and update it again on frame N + 2 when you replay. SaveState only records the state that the physics engine modifies during its update step. 

If you start adding / removing objects (e.g. bodies) during these frames, the SaveState function doesn't deal with that. If you added a body on frame N + 1,  you'll need to remove it when rewinding and then add it back on frame N + 1 again (with the proper initial position/velocity etc. because it won't be in the snapshot as the body didn't exist on frame N). These bodies will need to have the same BodyIDs, so this will work as long as you don't destroy/re-create the body, but only remove/add it to the physics system. If you wish to share state saved between server and client, you need to ensure that all APIs that modify the state of the world are called in the exact same order. So if the client creates physics objects for player 1 then 2 and the server creates the objects for 2 then 1 you already have a source of non-determinism (the body IDs will be different, which will influence the sorting of the constraints).

Determinism is incredibly hard to implement properly.

B.t.w. I can imagine that some API improvements can be made like a BodyInterface::CreateBodyWithProvidedID(...) function to get rid of the need to create objects in the same order. It's hard for me to say what you'll exactly run into.

**Reply by feixinyv** [reply]
Thanks very much!

**Comment by jrouwe** [comment]
I just submitted a BodyInterface::CreateBodyWithID function that should allow you to relax the creation order of bodies a bit and still have a deterministic simulation. See #247.

**Reply by feixinyv** [reply]
Wow, thanks a lot!
I am still doing the Prediction & Correction Roll-back Physics NetCode work, and I encountered the float determinstic problem in UE's logic code. I tried to change compile options of different platform, include /fp:precise, -fno-fast-math -ffp-contract=off, but it didn't work. 
I noticed that all vector calculation in Jolt are all SIMD instructions and disables FMADD instruction. Is that the key point to keep cross-platform determinstic of float? (Different compiler have different strategy to turn scalar calculation to SIMD)
What else should I care about to make float calculation determinstic?

**Reply by jrouwe** [reply]
As long as you compile in precise mode, the rules should be the same on all platforms (or at least the ones I tested). The biggest things I ran into were:

* FMADD = a + b * c, without the FMADD instruction you calculate b * c first, round it off to the nearest float and then add a which is slightly different because FMADD has higher intermediate precision
* If you use SIMD, take a good look at the order of operations e.g. an instruction may do (a + b) + (c + d) on one platform and ((a + b) + c) + d on another which gives slightly different results because of round off to float between every operation. If you do things in the right order and avoid instructions that have higher intermediate precision it doesn't matter if you use SIMD or scalar math.
* sinf, cosf, tanf etc. are approximations, they're implemented differently on different platforms so give different results
* std::sort is implemented differently across platforms, your elements will be sorted in a different way if they're 'equal'
 
I don't know if Unreal has any guarantees for cross-platform determinism, but if it hasn't then I think it would be a very big task to retro fit that.
